### PR TITLE
hubble: allow to filter agent events

### PR DESCRIPTION
**Pull request by tklauser**
This PR allows to apply filters on all event types and to filter agent events specifically. Previously (with #14168 in place), agent events were always returned leading to them being interleaved with flow events in Hubble CLI output.

See cilium/hubble#442 for the respective Hubble CLI PR, adding support for `hubble observe -t agent`.

### CODE REVIEW

**Comment by tklauser** [comment]
test-me-please

**Review by gandro — APPROVED** [review]
There is an overall API discussion to be had (e.g. should pod filters also work with endpoint agent events and such, and if we should revive the `GetEvents` proposal) but I think for the current implementation (where agent events are returned by `GetFlows`) this makes a ton of sense.

File: pkg/hubble/filters/event_type.go [hunk 1]
```diff
@@ -19,24 +19,34 @@ import (
...
+				}
+				if typeFilter.GetMatchSubType() && typeFilter.GetSubType() != event.SubType {
+					continue
+				}
+				return true
 			}
-
-			if typeFilter.GetMatchSubType() && typeFilter.GetSubType() != event.SubType {
-				continue
+		case *flowpb.AgentEvent:
```
Comment (gandro) [inline pkg/hubble/filters/event_type.go:42]: nit: Could you maybe add/extend a simple unit test ensuring that this case works as expected?

Comment (tklauser) [reply pkg/hubble/filters/event_type.go:42]: Added `TestEventTypeFilter` which should cover the basic cases.

**Comment by tklauser** [comment]
test-me-please

**Comment by tklauser** [comment]
retest-runtime

**Review by rolinh — COMMENTED** [review]
This looks pretty solid overall. There's one thing that I think needs attention (writing here as I can't comment on lines that are untouched by the PR): in `func newRingReader()` in `local_observer.go`, when rewinding, only flows are considered:
```go
		_, ok := e.Event.(*flowpb.Flow)
		if !ok || !filters.Apply(whitelist, blacklist, e) {
			continue
		}
```
I think the part of the code above needs to be updated such that filters are also applied on agent events. This means that with the current code, we're always rewinding the ring buffer all the way until the writer is met instead of stopping early based on the time range or filters. However, as @gandro mentioned in the comment above, we are planning to have an API discussion and this might not be relevant anymore anyway and is not too important as this is more an optimization than anything else.

**Comment by tklauser** [comment]
> I think the part of the code above needs to be updated such that filters are also applied on agent events. This means that with the current code, we're always rewinding the ring buffer all the way until the writer is met instead of stopping early based on the time range or filters. However, as @gandro mentioned in the comment above, we are planning to have an API discussion and this might not be relevant anymore anyway and is not too important as this is more an optimization than anything else.

Thanks @Rolinh for pointing that out. Given the CI run on this PR suceeded and in case there are no further review comments, I'd address this in a follow-up depending on the outcome of the API discussion.

**Review by kaworu — APPROVED** [review]

**Review by rolinh — APPROVED** [review]

**Comment by tklauser** [comment]
All required CI jobs passed (**K8s-1.20-kernel-4.9 ** was newly added while this PR was in flight). Marking as ready to merge.

**Pull request merged**